internal/target/repository_credential_library_test.go

```diff
@@ -0,0 +1,482 @@
+				return
+			}
+			require.NoError(err)
+			assert.Equal(tt.wantRowsDeleted, deletedRows)
+
+			// we should find the oplog for the delete of target credential libraries
+			err = db.TestVerifyOplog(t, rw, target.GetPublicId(), db.WithOperation(oplog.OpType_OP_TYPE_DELETE), db.WithCreateNotBefore(10*time.Second))
+			assert.NoError(err)
+		})
+	}
+}
+
+func TestRepository_SetTargetCredentialLibraries(t *testing.T) {
+	t.Parallel()
+	conn, _ := db.TestSetup(t, "postgres")
+	rw := db.New(conn)
+	wrapper := db.TestWrapper(t)
+	testKms := kms.TestKms(t, conn, wrapper)
+	repo, err := NewRepository(rw, rw, testKms)
+	require.NoError(t, err)
+
+	iamRepo := iam.TestRepo(t, conn, wrapper)
+	_, proj := iam.TestScopes(t, iamRepo)
+
+	cs := vault.TestCredentialStores(t, conn, wrapper, proj.GetPublicId(), 1)[0]
+	credLibs := vault.TestCredentialLibraries(t, conn, wrapper, cs.GetPublicId(), 2)
+	lib1 := credLibs[0]
+	lib2 := credLibs[1]
+
+	setupFn := func(target Target) []*CredentialLibrary {
+		credLibs := vault.TestCredentialLibraries(t, conn, wrapper, cs.GetPublicId(), 10)
+		clIds := make([]string, 0, len(credLibs))
+		for _, cl := range credLibs {
+			clIds = append(clIds, cl.PublicId)
+		}
+
+		_, _, created, err := repo.AddTargetCredentialLibraries(context.Background(), target.GetPublicId(), 1, clIds)
+		require.NoError(t, err)
+		require.Equal(t, 10, len(created))
+		return created
+	}
+	type args struct {
+		targetVersion uint32
+		clIds         []string
+		addToOrigLibs bool
+	}
+	tests := []struct {
+		name             string
+		setup            func(Target) []*CredentialLibrary
+		args             args
+		wantAffectedRows int
+		wantErr          bool
+		wantErrCode      errors.Code
+	}{
+		{
+			name:  "clear",
+			setup: setupFn,
+			args: args{
+				targetVersion: 2,
+				clIds:         []string{},
+			},
+			wantErr:          false,
+			wantAffectedRows: 10,
+		},
+		{
+			name:  "no-change",
+			setup: setupFn,
+			args: args{
+				targetVersion: 2,
+				clIds:         []string{},
+				addToOrigLibs: true,
+			},
+			wantErr:          false,
+			wantAffectedRows: 0,
+		},
+		{
+			name:  "add-sets",
+			setup: setupFn,
+			args: args{
+				targetVersion: 2,
+				clIds:         []string{lib1.PublicId, lib2.PublicId},
+				addToOrigLibs: true,
+			},
+			wantErr:          false,
+			wantAffectedRows: 2,
+		},
+		{
+			name:  "zero version",
+			setup: setupFn,
+			args: args{
+				targetVersion: 0,
+				clIds:         []string{lib1.PublicId, lib2.PublicId},
+				addToOrigLibs: true,
+			},
+			wantErr:     true,
+			wantErrCode: errors.InvalidParameter,
+		},
+		{
+			name:  "bad version",
+			setup: setupFn,
+			args: args{
+				targetVersion: 1000,
+				clIds:         []string{lib1.PublicId, lib2.PublicId},
+				addToOrigLibs: true,
+			},
+			wantErr:     true,
+			wantErrCode: errors.VersionMismatch,
+		},
+		{
+			name:  "remove existing and add users and grps",
```

Comment:
I think a few of these test names are left w/ copy/paste errors (referencing sets and users and groups).

Reply:
:/